invenio_records_resources/references/resolvers/base.py

```diff
@@ -0,0 +1,189 @@
+    responsibility of the EntityProxies created by the resolvers.
+    """
+
+    def _parse_ref_dict(self, ref_dict):
+        """Parse the referenced dict into a tuple (TYPE, ID)."""
+        return _parse_ref_dict(ref_dict)
+
+    def _parse_ref_dict_type(self, ref_dict):
+        """Parse the TYPE from the reference dict."""
+        return self._parse_ref_dict(ref_dict)[0]
+
+    def _parse_ref_dict_id(self, ref_dict):
+        """Parse the ID from the reference dict."""
+        return self._parse_ref_dict(ref_dict)[1]
+
+    def get_entity_proxy(self, ref_dict, check=True):
+        """Check compatibility and get a proxy for the referenced entity.
+
+        This method will optionally check the shape of the given reference
+        dict via ``matches_reference_dict()`` before creating a proxy via
+        ``_get_entity_proxy()``.
+        If this check fails, a ``ValueError`` will be raised.
+        """
+        if check and not self.matches_reference_dict(ref_dict):
+            raise ValueError(
+                f"{type(self).__name__} cannot handle "
+                f"the following reference dict: {ref_dict}"
+            )
+
+        return self._get_entity_proxy(ref_dict)
+
+    def reference_entity(self, entity, check=True):
+        """Check compatibility and create a reference dict for the entity.
+
+        This method will perform an optional instance check of the given
+        entity via ``matches_entity()`` before dumping a reference via
+        ``_reference_entity()``.
+        If this check fails, a ``ValueError`` will be raised.
+        """
+        if check and not self.matches_entity(entity):
+            raise ValueError(
+                f"{type(self).__name__} cannot handle "
+                f"the following entity: {entity}"
+            )
+
+        return self._reference_entity(entity)
+
+    @abstractmethod
+    def matches_reference_dict(self, ref_dict):
+        """Check if the ref_dict matches the expectations of this resolver."""
+        return False
+
+    @abstractmethod
+    def matches_entity(self, entity):
+        """Check if the entity matches the expectations of this resolver."""
+        return False
+
+    @abstractmethod
+    def _get_entity_proxy(self, ref_dict):
+        """The logic for building a proxy for the referenced entity.
+
+        Since the compatibility checks are already taken care of, this
+        method can assume that the data is valid and can focus on simply
+        creating the proxy for the referenced entity.
+        """
+        return None
+
+    @abstractmethod
+    def _reference_entity(self, entity):
+        """The logic building a reference dict from the given entity.
+
+        Since the compatibility checks are already taken care of, this method
+        can assume that the entity is compatible and can focus on simply
+        creating the reference dict for the given entity.
+        """
+        return None
```

Comment:
comment: I think those will be much clearer to me when concrete test cases highlight what they do.

Reply:
basically, the Resolver Registry is a translation layer between actual entities (e.g. User or Record objects) and reference dicts (e.g. `{"user": "1"}`, or `{"record": "abcd-1234"}`) the where you can dynamically register the handlers tasked with performing the actual translation (i.e. the Entity Resolvers)... it's similar to a dynamic factory pattern, i suppose.

the Entity Resolver interface is there to enable dumping and resolving of all kinds of entities with different structures.
for instance, User entities have their ID stored in an `.id` field, but for records it might be more appropriate to use the `recid` rather than their UUID (`.id`) for reference dicts.
so in this case, you can write a custom Entity Resolver and register it to the Resolver Registry, to enable dumping and resolving of these kinds of entities.

and the Entity Proxies are buffers in the "read"/"resolve" direction (ref-dict to entity), to postpone potentially heavy DB lookup operations until you *really* want to handle the referenced entity (`proxy_obj.resolve()`), mostly for performance reasons (cf. [owners in rdm-records](https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/records/systemfields/access/owners.py#L43))

Reply:
Make more sense now, thanks.